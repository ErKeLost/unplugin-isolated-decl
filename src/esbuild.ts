/**
 * This entry file is for esbuild plugin.
 *
 * @module
 */

import { IsolatedDecl } from './index'

/**
 * Esbuild plugin
 *
 * @example
 * ```ts
 * import { build } from 'esbuild'
 *
 * build({
 *   plugins: [require('unplugin-isolated-decl/esbuild')()],
 * })
 * ```
 */
export default IsolatedDecl.esbuild as typeof IsolatedDecl.esbuild
