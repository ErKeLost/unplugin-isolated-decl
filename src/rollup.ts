/**
 * This entry file is for Rollup plugin.
 *
 * @module
 */

import { IsolatedDecl } from './index'

/**
 * Rollup plugin
 *
 * @example
 * ```ts
 * // rollup.config.js
 * import IsolatedDecl from 'unplugin-isolated-decl/rollup'
 *
 * export default {
 *   plugins: [IsolatedDecl()],
 * }
 * ```
 */
export default IsolatedDecl.rollup as typeof IsolatedDecl.rollup
