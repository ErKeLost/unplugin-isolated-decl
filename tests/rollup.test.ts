import fs from 'node:fs/promises'
import path from 'node:path'
import { outputToSnapshot } from '@sxzz/test-utils'
import { rollup } from 'rollup'
import esbuild from 'rollup-plugin-esbuild'
import { describe, expect, test } from 'vitest'
import UnpluginIsolatedDecl from '../src/rollup'

describe('rollup', () => {
  const fixtures = path.resolve(__dirname, 'fixtures')
  const TEST_SANDBOX_FOLDER = path.resolve(__dirname, 'temp/rollup')

  test('generate basic', async () => {
    const input = path.resolve(__dirname, 'fixtures/basic/main.ts')
    const dist = path.resolve(__dirname, `${TEST_SANDBOX_FOLDER}/basic`)

    const bundle = await rollup({
      input,
      plugins: [
        UnpluginIsolatedDecl({
          extraOutdir: 'temp',
          autoAddExts: true,
        }),
        esbuild(),
      ],
      logLevel: 'silent',
    })

    const result = await bundle.generate({
      dir: dist,
    })

    expect(outputToSnapshot(result.output)).toMatchSnapshot()
  })

  test('write entry-points (#22)', async () => {
    const dir = 'entry-points-22'
    const input = {
      a: path.resolve(fixtures, dir, 'a/index.ts'),
      b: path.resolve(fixtures, dir, 'b/index.ts'),
    }
    const dist = path.resolve(TEST_SANDBOX_FOLDER, dir)

    const bundle = await rollup({
      input,
      plugins: [UnpluginIsolatedDecl(), esbuild()],
      logLevel: 'silent',
    })

    // https://github.com/unplugin/unplugin-isolated-decl/issues/22
    await bundle.write({
      dir: dist,
      entryFileNames: '[name]/index.js',
      preserveModules: true,
    })

    expect(await getFileSnapshot(dist)).toMatchSnapshot()
  })

  test('write entry-points (#34)', async () => {
    const dir = 'entry-points-34'
    const input = {
      index: path.resolve(fixtures, dir, 'index.ts'),
      bar: path.resolve(fixtures, dir, 'foo/bar/index.ts'),
    }
    const dist = path.resolve(TEST_SANDBOX_FOLDER, dir)

    const bundle = await rollup({
      input,
      plugins: [
        UnpluginIsolatedDecl({
          autoAddExts: true,
        }),
        esbuild(),
      ],
      logLevel: 'silent',
    })

    await bundle.write({ dir: dist })

    expect(await getFileSnapshot(dist)).toMatchSnapshot()
  })
})

async function getFileSnapshot(dir: string) {
  /**
   * Map written output from file system rather than from bundle due to
   * module execution order not consistent
   *
   * @see https://github.com/rollup/rollup/issues/3888
   */
  const files = (
    await fs.readdir(dir, { recursive: true, withFileTypes: true })
  ).filter((it) => it.isFile())

  const snapshot = await Promise.all(
    files.map(async (it) => {
      const absolute = path.resolve(it.parentPath, it.name)
      const filePath = path.relative(dir, absolute)
      const content = await fs.readFile(absolute, 'utf-8')

      return `// ${filePath.replaceAll('\\', '/')}\n${content.toString()}`
    }),
  )
  return snapshot
}
