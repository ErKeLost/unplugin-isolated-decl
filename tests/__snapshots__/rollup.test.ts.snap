// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rollup > generate basic 1`] = `
"// main.js
function Component() {
  return /* @__PURE__ */ React.createElement("div", null, "I'm a div in a tsx component!");
}

function hello(s) {
  return "hello" + s;
}
let c = Component;
let num = 1;

export { c, hello, num };

// temp/component.d.ts
export declare function Component(): React.JSX.Element;

// temp/main.d.ts
import { type Num } from './types.js';
export type Str = string;
export declare function hello(s: Str): Str;
export declare let c: React.JSX.Element;
export declare let num: Num;

// temp/types.d.ts
import type { Num2 } from './types2.js';
export type Num = Num2;

// temp/types2.d.ts
export type Num2 = number;
"
`;

exports[`rollup > write entry-points (#22) 1`] = `
[
  "// b/index.d.ts
export { ComponentB } from "./ComponentB/ComponentB";
",
  "// b/index.js
export { ComponentB } from './ComponentB/ComponentB/index.js';
",
  "// b/ComponentB/ComponentB/index.d.ts
import React from "react";
interface ComponentBProps {
    test: string;
}
export declare function ComponentB(props: ComponentBProps): React.JSX.Element;
export {};
",
  "// b/ComponentB/ComponentB/index.js
import React from 'react';

function ComponentB(props) {
  const { test } = props;
  return /* @__PURE__ */ React.createElement(React.Fragment, null, "ComponentB: ", test);
}

export { ComponentB };
",
  "// a/index.d.ts
export { ComponentA } from "./ComponentA/ComponentA";
",
  "// a/index.js
export { ComponentA } from './ComponentA/ComponentA/index.js';
",
  "// a/ComponentA/ComponentA/index.d.ts
import React from "react";
interface ComponentAProps {
    test: string;
}
export declare function ComponentA(props: ComponentAProps): React.JSX.Element;
export {};
",
  "// a/ComponentA/ComponentA/index.js
import React from 'react';

function ComponentA(props) {
  const { test } = props;
  return /* @__PURE__ */ React.createElement(React.Fragment, null, "ComponentA: ", test);
}

export { ComponentA };
",
]
`;

exports[`rollup > write entry-points (#34) 1`] = `
[
  "// bar.d.ts
export * as baz from './foo/bar/baz.js';
export type CustomType = 'unimportant' | 'only' | 'showcase';
",
  "// bar.js
export { b as baz } from './index-B_ebAVVP.js';
",
  "// index-B_ebAVVP.js
function important() {
  console.log("very! important function");
}

var baz = /*#__PURE__*/Object.freeze({
  __proto__: null,
  important: important
});

export { baz as b, important as i };
",
  "// index.d.ts
export declare function foo(): void;
export * from './foo/bar/baz.js';
",
  "// index.js
export { i as important } from './index-B_ebAVVP.js';

function foo() {
  console.log("important function");
}

export { foo };
",
  "// foo/bar/baz.d.ts
export declare function important(): void;
",
]
`;
